examples/simulation_pytorch/dataset.py

```diff
@@ -0,0 +1,75 @@
+"""Partitioned version of CIFAR-10 dataset."""
+
+from typing import List, Tuple, cast
+
+import numpy as np
+import torch
+import torchvision
+import torchvision.transforms as transforms
+from torch.utils.data import DataLoader, TensorDataset
+
+XY = Tuple[np.ndarray, np.ndarray]
+XYList = List[XY]
+PartitionedDataset = List[Tuple[XY, XY]]
+
+
+def cifar_to_numpy():
+    """Download dataset from torchvision and convert it to numpy array."""
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
+    )
+
+    trainset = torchvision.datasets.CIFAR10(
+        root="./data", train=True, download=True, transform=transform
+    )
+    trainloader = torch.utils.data.DataLoader(
+        trainset, batch_size=len(trainset), shuffle=True, num_workers=2
+    )
+
+    testset = torchvision.datasets.CIFAR10(
+        root="./data", train=False, download=True, transform=transform
+    )
+    testloader = torch.utils.data.DataLoader(
+        testset, batch_size=len(testset), shuffle=False, num_workers=2
+    )
+
+    train_X = next(iter(trainloader))[0].numpy()
+    test_X = next(iter(testloader))[0].numpy()
+    train_Y = next(iter(trainloader))[1].numpy()
+    test_Y = next(iter(testloader))[1].numpy()
+    return (train_X, train_Y), (test_X, test_Y)
```

Comment:
The model was not converging so I tested a bit extensively. The bug was `next(iter(trainloader))` was called for x and y separately.

This fixes this issue:

```python
    train_xy = next(iter(trainloader))
    test_xy = next(iter(testloader))

    train_x = train_xy[0].numpy()
    train_y = train_xy[1].numpy()

    test_x = test_xy[0].numpy()
    test_y = test_xy[1].numpy()

    return (train_x, train_y), (test_x, test_y)
```

While debugging I also wrote a more PyTorch-only version without NumPy keeping close to your version. I would suggest we take the PyTorch way with `random_splits` as it also loads faster.

This would be the alternative version of `dataset.py`

```python
"""Partitioned version of CIFAR-10 dataset."""

from typing import List, Tuple

import torch
import torchvision
import torchvision.transforms as transforms


def cifar10() -> Tuple[torchvision.datasets.CIFAR10, torchvision.datasets.CIFAR10]:
    """Download dataset from torchvision and convert it to numpy array."""
    transform = transforms.Compose(
        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
    )

    trainset = torchvision.datasets.CIFAR10(
        root="./data", train=True, download=True, transform=transform
    )
    testset = torchvision.datasets.CIFAR10(
        root="./data", train=False, download=True, transform=transform
    )

    return trainset, testset


def create_partitioned_dataloaders(dataset, num_partitions: int) -> List[torch.utils.data.DataLoader]:
    """Create partitioned version of a source dataset."""
    num_per_partition = len(dataset)//num_partitions
    # Seed generator for predictable shuffling of dataset
    generator = torch.Generator().manual_seed(42)
    dataset_partition = torch.utils.data.random_split(dataset, [
        num_per_partition] * num_partitions, generator=generator)
    dataloaders = [torch.utils.data.DataLoader(
        ds_p, batch_size=32, shuffle=True) for ds_p in dataset_partition]
    return dataloaders


def load(num_partitions: int) -> List[Tuple[torch.utils.data.DataLoader, torch.utils.data.DataLoader]]:
    """Create partitioned version of CIFAR-10."""
    trainset, testset = cifar10()
    trainloaders = create_partitioned_dataloaders(trainset, num_partitions)
    testloaders = create_partitioned_dataloaders(testset, num_partitions)
    list_of_dataloaders = [(train_dl, test_dl) for train_dl,
                           test_dl in zip(trainloaders, testloaders)]
    return list_of_dataloaders
```

Reply:
@tanertopal Thanks a lot for finding the bug. I am sorry I might have overlooked this part.

Regarding the pytorch only solution, it is a creative idea and  I really liked it. One more thing, I found this issue with random split [open issue](https://github.com/PyTorchLightning/lightning-bolts/issues/312) which says the split doesn't assure balanced class labels. Having said that it is not a significant problem for us as we expect Non-IID distribution anyways.

Reply:
@MakGulati Cool that you found the issue.  It should be fine as in the examples, it's not the goal to provide ideal solutions for targeted data distributions but rather give a starting point for the user. I realized you can further simplify the dataset loading in your version with:

```python
def cifar10():
    """Download dataset from torchvision and convert it to numpy array."""
    transform = transforms.Compose(
        [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
    )

    trainset = torchvision.datasets.CIFAR10(
        root="./data", train=True, download=True, transform=transform
    )
    testset = torchvision.datasets.CIFAR10(
        root="./data", train=False, download=True, transform=transform
    )
 
    train_xy = trainset.data, np.array(trainset.targets)
    test_xy = testset.data, np.array(testset.targets)

    return train_xy, test_xy
```

I am by no mean a PyTorch expert so not sure if one would use it that way.

Feel free to commit your preferred solution. If you go with the NumPy solution you will have to shuffle the dataset in the create_partitions function as in: https://github.com/adap/flower/blob/main/examples/simulation/dataset.py#L44

Reply:
@tanertopal I think it should be done this time. Please let me know otherwise. I learnt a lot of things from this PR. :)

Reply:
Happy to hear!